bplustree_test.go

```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
+		maxElems: 10, evaluator: func(k1 Key, k2 Key) bool {
+			result := k1.Compare(k2)
+			if result < 10 && result > -10 {
+				return true
+			}
+
+			return false
+		},
+	}
+
+	result, err := newtree.Search(ss)
+	if err != nil {
+		t.Errorf("Expected to find the key: %v", ss.searchKey)
+		t.FailNow()
+	}
+	if result == nil {
+		t.Errorf("Expected to get result but nil: %v", result)
+		t.FailNow()
+	}
+
+	if len(result) < ss.maxElems || len(result) > ss.maxElems+1 {
+		t.Errorf("Expected to find atleast %d results, got: %d", ss.maxElems, len(result))
+		t.FailNow()
+	}
+
+	// Note that our values are +10 of the key.
+	startKey := makeTestingKeyStr(strHash(testingKey) - 5)
+	expectedVal := int(strHash(startKey) + 10)
+	for i := 0; i < len(result); i++ {
+		//fmt.Println(reflect.TypeOf(result[i]))
+		actualVal := result[i].(*TestingElemStr).val
+		if actualVal != expectedVal {
+			t.Errorf("At index %d: expected: %d, got %v", i, expectedVal, actualVal)
+			t.FailNow()
+		}
+		// and increment by 1.
+		expectedVal++
+	}
+}
+
+func TestSearchStr7(t *testing.T) {
+
+	// case 7. use of evaluator with maxElems = 10, such that
+	// evaluator yields less than 10 elements. We should only
+	// get less than 10 elements.
+
+	testingKey := makeTestingKeyStr(100)
+	ss := SearchSpecifier{searchKey: testingKey, direction: Both,
+		maxElems: 10, evaluator: func(k1 Key, k2 Key) bool {
+			result := k1.Compare(k2)
+			if result < 3 && result > -3 {
+				return true
+			}
+
+			return false
+		},
+	}
+
+	result, err := newtree.Search(ss)
+	if err != nil {
+		t.Errorf("Expected to find the key: %v", ss.searchKey)
+		t.FailNow()
+	}
+	if result == nil {
+		t.Errorf("Expected to get result but nil: %v", result)
+		t.FailNow()
+	}
+
+	if len(result) != 5 {
+		t.Errorf("Expected to find 5 results, got: %d", len(result))
+		t.FailNow()
+	}
+
+	// Note that our values are +10 of the key.
+	startKey := makeTestingKeyStr(strHash(testingKey) - 2)
+	expectedVal := int(strHash(startKey) + 10)
+	for i := 0; i < len(result); i++ {
+		//fmt.Println(reflect.TypeOf(result[i]))
+		actualVal := result[i].(*TestingElemStr).val
+		if actualVal != expectedVal {
+			t.Errorf("At index %d: expected: %d, got %v", i, expectedVal, actualVal)
+			t.FailNow()
+		}
+		// and increment by 1.
+		expectedVal++
+	}
+}
+
+func TestRemoveStr1(t *testing.T) {
+	fmt.Println("Starting remove in order tests")
+	for i := 0; i < numElems; i++ {
+		key := makeTestingKeyStr(i)
+		fmt.Printf("removing key: %v", key)
+		err := newtree.Remove(key)
+		if err != nil {
+			t.Errorf("Failed to remove key: %v got %v", key, err)
+			t.FailNow()
+		}
+	}
+}
+
+func TestRemoveStr2(t *testing.T) {
+	TestInsertStr(t)
+	fmt.Println("Starting remove in reverse order tests")
+	for i := numElems - 1; i >= 0; i-- {
+		key := makeTestingKeyStr(i)
+		fmt.Printf("removing key: %v", key)
+		err := newtree.Remove(key)
+		if err != nil {
+			t.Errorf("Failed to remove key: %v got %v", key, err)
+			t.FailNow()
+		}
+	}
```

Comment:
do lookup for some elements to make sure nothing is left. also would be useful to do some lookups of existing and deleted keys during the remove to make sure lookups continue to work

Reply:
sure.